File: leetcode/valid-palindrome.java
// https://leetcode.com/problems/valid-palindrome/

class Solution {
    public boolean isPalindrome(String s) {
        char[] input = s.toCharArray();
        int left = 0;
        int right = input.length-1;
        
        while(left<right){
            while(!Character.isLetterOrDigit(input[left]) && left < right) left++;
            while(!Character.isLetterOrDigit(input[right]) && left < right) right--;
            if(!(""+input[left++]).equalsIgnoreCase(""+input[right--])) return false;
        }
        return true;
    }
}

File: leetcode/valid_palindrome.py
# https://leetcode.com/problems/valid-palindrome/description/


class Solution:
    def isPalindrome(self, s: str) -> bool:
        left: int = 0
        right: int = len(s) - 1

        while left < right:
            while not s[left].isalpha() and not s[left].isnumeric() and left < right:
                left += 1
            while not s[right].isalpha() and not s[right].isnumeric() and left < right:
                right -= 1
            if left < right:
                if s[left].lower() != s[right].lower():
                    return False
                left += 1
                right -= 1
        return True
